Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp

```diff
@@ -119,72 +152,6 @@ CTFontRef FontPlatformData::registeredFont() const
     return nullptr;
 }
 
-inline int mapFontWidthVariantToCTFeatureSelector(FontWidthVariant variant)
-{
-    switch (variant) {
-    case FontWidthVariant::RegularWidth:
-        return kProportionalTextSelector;
-
-    case FontWidthVariant::HalfWidth:
-        return kHalfWidthTextSelector;
-
-    case FontWidthVariant::ThirdWidth:
-        return kThirdWidthTextSelector;
-
-    case FontWidthVariant::QuarterWidth:
-        return kQuarterWidthTextSelector;
-    }
-
-    ASSERT_NOT_REACHED();
-    return kProportionalTextSelector;
-}
-
-static RetainPtr<CFDictionaryRef> cascadeToLastResortAttributesDictionary()
-{
-    auto lastResort = adoptCF(CTFontDescriptorCreateWithNameAndSize(CFSTR("LastResort"), 0));
```

Comment:
Is it right to no longer create the font with the "LastResort" postscript name?

If so, shouldn't we remove any mentions to it on code?

For example, we have special handling for it here: https://searchfox.org/wubkat/source/Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm#254

Reply:
Actually we are not creating a font with such a name. What we do is:

1. `CTFontDescriptorCreateWithNameAndSize` creates a font descriptor with `LastResort` postscript name
2. creates a dictionary that maps from `kCTFontCascadeListAttribute` attribute to an array with a single descriptor, the `LastResort` descriptor in [1]
3. `CTFontDescriptorCreateWithAttributes` uses the dictionary in [2] to create a font descriptor that has `kCTFontCascadeListAttribute` attribute set to a list with a single descriptor, `LastResort` descriptor.
4. `FontPlatformData::ctFont()` will create a new `CTFont` that is a copy of the existing `CTFont` held at `m_font` but now with its `kCTFontCascadeListAttribute` pointing to a list with a single font descriptor, the `LastResort` descriptor
5. if the requested `FontWidthVariant` is not `::RegularWidth` we will create a new font descriptor based on the `CTFont` at [4], and then a new `CTFont` based on such descriptor. We will hold such font at `m_ctFont` and this is used at some points in our code base, until this patch. I don't really understand why we need this, but it seems that after this patch, the only difference is that the `CTFont` we return will no longer have its `kCTFontCascadeListAttribute` attribute pointing to a list containing the `LastResort` font descriptor.

Reply:
> If so, shouldn't we remove any mentions to it on code?

No, if we run into LastResort it means that CoreText couldn't find a font that can render the glyph, and we do want to continue to have special handling for that case.

Reply:
Ok, so the question we have to answer here is why did we need a `CTFont` that  has its `kCTFontCascadeListAttribute` set to a list containing a single descriptor (the one from `LastResort`).

Reply:
The `FontWidthVariant` stuff is used for implementing `text-combine-horizontal` in https://searchfox.org/wubkat/source/Source/WebCore/rendering/RenderCombineText.cpp#139
AFAICT applying it to both m_font and m_ctFont would have been the right thing to do; we're just lucky that the various places where we looked up m_font instead of m_ctFont would (in any give font) return the same values.
I checked the relevant `text-combine-horizontal` effects, and they work as expected.

Reply:
> No, if we run into LastResort it means that CoreText couldn't find a font that can render the glyph, and we do want to continue to have special handling for that case.

(Although to be honest, I have no idea how to run into this case. We don't seem to be hitting it when we render unknown codepoints.)

Reply:
> Ok, so the question we have to answer here is why did we need a CTFont that has its kCTFontCascadeListAttribute set to a list containing a single descriptor (the one from LastResort).

According to Myles, it was because we used to check for LastResort in order to know whether a font needed fallback in complex text. But our fallback behavior seems to work just fine through multiple fonts with the patch enabled. (We do have a bug when we hit a generic keyword, but that's a pre-existing bug.)

Reply:
I guess to make sure we would need a test in which we use the `CodePath::Complex` path (we can use specific codepoints, for example) and use a `font-family` list in which the first font does not support a code point but the second does, and then we use as `-ref` the text but with a `font-family` list composed only by the second font in the test's `font-family` list. We can use Ahem as the second font to make sure it does not coincide with a font that CT would fallback to.

Reply:
Added fast/text/font-fallback.html